File: examples/AgenticPatchRunner.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

public sealed record ToolCall(string Name, string Path, IReadOnlyDictionary<string, object?> Arguments);

public sealed record Observation(bool Succeeded, string Output);

public interface IWorkspace
{
    string Revision { get; }
    Task<Observation> ExecuteAsync(ToolCall call);
}

public interface IModel
{
    Task<ToolCall> NextCallAsync(string goal, IReadOnlyList<(ToolCall Call, Observation Observation)> transcript);
}

public sealed class AgenticPatchRunner
{
    private readonly IWorkspace workspace;
    private readonly IModel model;
    private readonly TimeSpan turnTimeout;

    public AgenticPatchRunner(IWorkspace workspace, IModel model, TimeSpan turnTimeout)
    {
        this.workspace = workspace;
        this.model = model;
        this.turnTimeout = turnTimeout;
    }

    /* @agent-context design
     * Ask the model for one tool call per turn. The original batched protocol kept executing stale assumptions after its first edit changed the workspace, whereas single-call turns provide exact approval, cancellation, and checkpoint boundaries.
     */
    public async Task<string> RunAsync(string goal, bool allowWrites)
    {
        var transcript = new List<(ToolCall Call, Observation Observation)>();
        var expectedRevision = workspace.Revision;

        while (true)
        {
            var startedAt = DateTimeOffset.UtcNow;
            var call = Normalize(await model.NextCallAsync(goal, transcript));

            /* @agent-context invariant
             * Authorization must inspect the normalized call and the executor must receive that same value. Checking the raw proposal first would allow path cleanup or alias expansion to change the operation after approval.
             */
            if (call.Name == "apply_patch" && !allowWrites)
            {
                return "blocked: workspace writes were not authorized";
            }

            /* @agent-context dependency
             * The editor bridge uses revisions to represent human activity, including edits that are later undone to identical bytes. A content hash cannot preserve that signal and could allow an agent to overwrite a human rejection.
             */
            if (workspace.Revision != expectedRevision)
            {
                return "blocked: workspace changed during the run";
            }

            var observation = await workspace.ExecuteAsync(call);
            transcript.Add((call, observation));
            expectedRevision = workspace.Revision;

            /* @agent-context history
             * Retain failed observations in the transcript. Incident AGENT-42 dropped executor errors and retried an invalid rename until the tool budget expired, while replay logs omitted the evidence needed to diagnose the loop.
             */
            if (observation.Succeeded && call.Name == "finish")
            {
                return observation.Output;
            }

            if (DateTimeOffset.UtcNow - startedAt > turnTimeout)
            {
                return "blocked: model turn exceeded its time budget";
            }
        }
    }

    /* @agent-context compatibility
     * Persist prompt paths with forward slashes so evaluation transcripts replay identically across operating systems. Native separator conversion belongs in the workspace implementation immediately before I/O.
     */
    private static ToolCall Normalize(ToolCall call)
    {
        return call with { Path = call.Path.Replace('\\', '/') };
    }
}

File: examples/AgenticPatchRunner.ts
type ToolName = "read_file" | "apply_patch" | "run_tests" | "finish";

interface ToolCall {
  name: ToolName;
  path: string;
  arguments: Readonly<Record<string, unknown>>;
}

interface Observation {
  succeeded: boolean;
  output: string;
}

interface Workspace {
  readonly revision: string;
  execute(call: ToolCall): Promise<Observation>;
}

interface Model {
  nextCall(goal: string, transcript: ReadonlyArray<readonly [ToolCall, Observation]>): Promise<ToolCall>;
}

export class AgenticPatchRunner {
  constructor(
    private readonly workspace: Workspace,
    private readonly model: Model,
    private readonly turnTimeoutMs = 30_000
  ) {}

  /* @agent-context design
   * Request one tool call per model turn. Earlier versions accepted batches, but the remaining calls became stale as soon as the first edit changed the workspace. A single-call loop also gives cancellation and approval checks an exact boundary.
   */
  async run(goal: string, allowWrites: boolean): Promise<string> {
    const transcript: Array<readonly [ToolCall, Observation]> = [];
    let expectedRevision = this.workspace.revision;

    while (true) {
      const startedAt = performance.now();
      const call = this.normalize(await this.model.nextCall(goal, transcript));

      /* @agent-context invariant
       * Authorization applies to the normalized call that reaches the executor. Never authorize the raw model proposal and rewrite it afterward because aliases and relative segments can change the target or convert an apparent read into a write.
       */
      if (call.name === "apply_patch" && !allowWrites) {
        return "blocked: workspace writes were not authorized";
      }

      /* @agent-context dependency
       * The editor bridge increments revisions for human activity, including edit-and-undo sequences that restore identical bytes. Replacing this check with a content hash would erase the signal that a human reconsidered the affected code.
       */
      if (this.workspace.revision !== expectedRevision) {
        return "blocked: workspace changed during the run";
      }

      const observation = await this.workspace.execute(call);
      transcript.push([call, observation]);
      expectedRevision = this.workspace.revision;

      /* @agent-context history
       * Preserve failed observations in the transcript. Incident AGENT-42 repeatedly retried an invalid rename because the failure was discarded before the following model turn, exhausting the tool budget and obscuring the original cause in replay diagnostics.
       */
      if (observation.succeeded && call.name === "finish") {
        return observation.output;
      }

      if (performance.now() - startedAt > this.turnTimeoutMs) {
        return "blocked: model turn exceeded its time budget";
      }
    }
  }

  /* @agent-context compatibility
   * Persist prompt paths with POSIX separators so recorded evaluations replay consistently on Windows, macOS, and Linux. Native path conversion belongs in the workspace adapter immediately before filesystem access.
   */
  private normalize(call: ToolCall): ToolCall {
    return { ...call, path: call.path.replaceAll("\\", "/") };
  }
}
